### PR TITLE
Solve #721 Deberta masklm model

### PR DESCRIPTION
**Pull request by Plutone11011**
PR for DeBERTa mask model and preprocessor.

Related to this, I noticed that DeBERTaV3 uses a different technique for masking, however the backbone explicitly doesn't include it so I don't know if it is relevant here.

### CODE REVIEW

**Comment by mattdangerw** [comment]
@Plutone11011 thanks! Will take a pass soon.

> I noticed that DeBERTaV3 uses a different technique for masking

Can you elaborate? If it's all preprocessing, like [this comment](https://github.com/keras-team/keras-nlp/pull/725#discussion_r1097309749), I think it is fine to cover the fancier token masking schemes down the road.

If this is a whole different setup, or different "head" on the backbone, we might want to consider some changes.

**Comment by Plutone11011** [comment]
> @Plutone11011 thanks! Will take a pass soon.
> 
> > I noticed that DeBERTaV3 uses a different technique for masking
> 
> Can you elaborate? If it's all preprocessing, like [this comment](https://github.com/keras-team/keras-nlp/pull/725#discussion_r1097309749), I think it is fine to cover the fancier token masking schemes down the road.
> 
> If this is a whole different setup, or different "head" on the backbone, we might want to consider some changes.

In https://arxiv.org/pdf/2111.09543.pdf the authors talk about a replacement for MLM called RTD (replaced token detection), this is what mainly distinguishes DeBERTaV3 from previous versions. It is still somewhat based on MLM, but it uses a GAN style approach, jointly training a MLM generator and a discriminator/classifier. See section Section 3.1 and 2.3.2 of the paper, specifically.

**Comment by mattdangerw** [comment]
Ah right! DeBERTaV3 uses electra style (GAN like) pre-training.

I think it is still totally valid to ship a mlm task for deberta, but we can probably make a note in the docstring that this is not the "task setup" used by deberta during pre-training.

File: keras_nlp/models/deberta_v3/deberta_v3_masked_lm.py
Comment (mattdangerw) [inline keras_nlp/models/deberta_v3/deberta_v3_masked_lm.py:51]: switch this to the deberta repo

File: keras_nlp/models/deberta_v3/deberta_v3_masked_lm.py
Comment (mattdangerw) [inline keras_nlp/models/deberta_v3/deberta_v3_masked_lm.py:129]: We should probably use the same "approximate gelu" as the backbone itself here-> https://github.com/keras-team/keras-nlp/blob/4c1c6ae9e5a3adcf80271ba206b6835caf1b39f2/keras_nlp/models/deberta_v3/deberta_v3_backbone.py#L155

File: keras_nlp/models/deberta_v3/deberta_v3_masked_lm_preprocessor.py
Comment (mattdangerw) [inline keras_nlp/models/deberta_v3/deberta_v3_masked_lm_preprocessor.py:76]: format this to fit our lint limit

```
preprocessor = keras_nlp.models.DebertaV3MaskedLMPreprocessor.from_preset(
    "deberta_v3_base_en",
)
```

File: keras_nlp/models/deberta_v3/deberta_v3_tokenizer.py
Comment (mattdangerw) [inline keras_nlp/models/deberta_v3/deberta_v3_tokenizer.py:76]: I think we should add a check for the mask token here, which might also mean you need to update some unit tests for the preprocessor and tokenizer layers (so they add a mask token to the vocabulary).

File: keras_nlp/models/deberta_v3/deberta_v3_masked_lm_test.py
Comment (mattdangerw) [inline keras_nlp/models/deberta_v3/deberta_v3_masked_lm_test.py:59]: nice! should this not be a list? is it just a comma separate string?

File: keras_nlp/models/deberta_v3/deberta_v3_masked_lm_test.py
Comment (mattdangerw) [inline keras_nlp/models/deberta_v3/deberta_v3_masked_lm_test.py:64]: I don't think this comment applies here.

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks! This looks great! Left a few comments

File: keras_nlp/models/deberta_v3/deberta_v3_masked_lm_test.py
Comment (Plutone11011) [reply keras_nlp/models/deberta_v3/deberta_v3_masked_lm_test.py:59]: yes it seems to be a comma separated string. 
From the options [doc](https://github.com/google/sentencepiece/blob/master/doc/options.md)
`--user_defined_symbols (comma separated list of user defined symbols)  type: std::string default: ""`

File: keras_nlp/models/deberta_v3/deberta_v3_masked_lm_preprocessor.py
Comment (Plutone11011) [reply keras_nlp/models/deberta_v3/deberta_v3_masked_lm_preprocessor.py:76]: The format script is not detecting it, is it because it's docstrings?

**Comment by Plutone11011** [comment]
Addressed the comments, also added mask check in tokenizer and I had to adjust vocab size and some tests as a result

File: keras_nlp/models/deberta_v3/deberta_v3_tokenizer.py
Comment (Plutone11011) [reply keras_nlp/models/deberta_v3/deberta_v3_tokenizer.py]: So, I added the check and as you can see it fails the preset tests, because apparently there isn't a `[MASK]` token in the `deberta_v3_extra_small_en` vocabulary, which seems strange to me (I've tried with `deberta_v3_base_en` too). Any advice?

**Comment by mattdangerw** [comment]
Everything looking good here, we can just merge this with #759 when it is ready. We can't merge before without breaking the main model usages.

**Comment by mattdangerw** [comment]
Ok #759 is merged, so we should be able to rebase this and get things working. You can use the colab I linked above to test things out (I recommend a GPU runtime).

**Comment by mattdangerw** [comment]
@Plutone11011 is this ready for review again? If so can take a pass tomorrow.

**Comment by Plutone11011** [comment]
> @Plutone11011 is this ready for review again? If so can take a pass tomorrow.

Yes, I've checked the notebook, training and preprocessor work, there is however still a problem when calling `detokenize` that I haven't delved into, it doesn't find the `[MASK]` id.

**Comment by Plutone11011** [comment]
@mattdangerw in the colab notebook the call to `detokenize` yields an `OutOfRangeError, invalid id 128000', basically unable to find the `[MASK]` id. I haven't followed closely #759 but I guess this is due to the fact that the `[MASK]` token is handled internally by `DebertaV3Tokenizer` in `self.mask_token_id`, whereas `detokenize` calls the SentencePiece method. Do you think we should implement a `detokenize` method inside `DebertaV3Tokenizer`?

**Comment by mattdangerw** [comment]
@Plutone11011 thanks for checking this out! IMO we don't need to block on the detokenize functionality (it is not critical to any MLM workflow), but it could be worth handling that in a follow up.

I'll take a pass over the code again shortly.

**Review by mattdangerw — APPROVED** [review]
Thank you! This was a bit of a journey! All looks good to me.

**Pull request merged**
